src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -899,13 +899,25 @@ void sort()
             }
             else if (count <= F)
             {
-                // Happy path. We are not late or (for levelled) we are only so late that a compaction now will have the
-                // same effect as doing levelled compactions one by one. Compact all.
+                /**
+                 * Happy path. We are not late or (for levelled) we are only so late that a compaction now will
+                 * have the  same effect as doing levelled compactions one by one. Compact all. We do not cap
+                 * this pick at maxSSTablesToCompact due to an assumption that maxSSTablesToCompact is much
+                 * greater than F. See {@link Controller#MAX_SSTABLES_TO_COMPACT_OPTION} for more details.
+                 */
                 selected = CompactionPick.create(index, liveSet);
             }
+            else if (count <= F * controller.getF(index + 1))
+            {
+                // Compaction is a bit late, but not enough to jump levels via layout compactions. We need a special
+                // case to cap compaction pick at maxSSTablesToCompact.
+                selected = CompactionPick.create(index, liveSet.subList(0, Math.min(maxSSTablesToCompact, count)));
+                if (count > maxSSTablesToCompact)
+                    pending = ImmutableList.of(CompactionPick.create(index, liveSet.subList(maxSSTablesToCompact, count)));
```

Comment:
Can these be more than one compaction (e.g. if max is F and we have e.g. 3*F sstables)?

Reply:
Fixed in 5eb54c4fffe72cd4e7a7f322eb692f688c59320b.